java/core/src/java/org/apache/orc/impl/ReaderImpl.java

```diff
@@ -569,7 +561,13 @@ public ReaderImpl(Path path, OrcFile.ReaderOptions options) throws IOException {
       this.rowIndexStride = tail.getFooter().getRowIndexStride();
       this.contentLength = tail.getFooter().getContentLength();
       this.numberOfRows = tail.getFooter().getNumberOfRows();
-      this.userMetadata = tail.getFooter().getMetadataList();
+      this.userMetadata = tail.getFooter().getMetadataList()
+              .stream().collect(Collectors.toMap(
```

Comment:
If you don't mind, can we indent with `4-space` instead of `8-space`?

Reply:
The java stream apis are very slow. It would perform better if you used a loop. This is especially important, because now the set is build preemptively, even if the user doesn't need the metadata values.

Reply:
> If you don't mind, can we indent with `4-space` instead of `8-space`?
I will fix

Reply:
> The java stream apis are very slow. It would perform better if you used a loop. This is especially important, because now the set is build preemptively, even if the user doesn't need the metadata values.

How about lazy initialization of the set when the getMetadataValue and hasMetadataValue methods need to be called

Reply:
> Thank you for making a PR, @guiyanakuang . Do you think we can have a benchmark result for your claim?
> 
> > There may be performance problems if there are more userMetadata keys
> 
> cc @pgaref , @williamhyun

Sorry for the late reply due to different time zones. I made a guess based on the time complexity of the algorithm and have not yet benchmarked it. getMetadataValue is used in compatibility checking and metadata merging, which would have an O(n^2) time complexity if used in a loop.